### PR TITLE
issues#1946 add [Druid] Source & Sink

### PR DESCRIPTION
**Pull request by guanboo**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by guanboo** [comment]
fix:issues#1946 add [Druid] Source & Sink

**Comment by guanboo** [comment]
fix:issues#1946 add [Druid] Source & Sink

File: seatunnel-connectors-v2/connector-druid/pom.xml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-druid/pom.xml:18]: add license header

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
+package org.apache.seatunnel.connectors.seatunnel.druid.client;
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:18]: same above

**Review by ashulin — COMMENTED** [review]
1. Please add the license header.
2. Maybe you can add druid dialect in jdbc connector.

**Comment by ashulin** [comment]
@ic4y PTAL, thx.

File: seatunnel-connectors-v2/connector-druid/pom.xml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        <dependency>
+            <groupId>org.apache.druid</groupId>
+            <artifactId>druid-core</artifactId>
+            <version>RELEASE</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.druid</groupId>
+            <artifactId>druid-indexing-service</artifactId>
+            <scope>compile</scope>
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-druid/pom.xml]: We already have the class JsonUtil, please use this.

File: seatunnel-connectors-v2/connector-druid/pom.xml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            <groupId>org.apache.calcite.avatica</groupId>
+            <artifactId>avatica-core</artifactId>
+            <version>1.15.0</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.druid</groupId>
+            <artifactId>druid-core</artifactId>
+            <version>RELEASE</version>
+            <scope>compile</scope>
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-druid/pom.xml]: The dependency version should manage by root pom

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        ArrayList<String> fieldNames = new ArrayList<>();
+
+        try {
+            ResultSetMetaData resultSetMetaData = getResultSetMetaData();
+            for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+                fieldNames.add(resultSetMetaData.getColumnName(i));
+                seaTunnelDataTypes.add(DruidTypeMapper.informationMapping.get(resultSetMetaData.getColumnTypeName(i)));
+            }
+        } catch (Exception e) {
+            LOGGER.warn("get row type info exception", e);
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java]: Should throw exception?

**Review by hailin0 — COMMENTED** [review]
Can you add connector-v2 e2e-testcase & docs

reference：#2499

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * guanbo
+ */
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java]: delete meaningless comments

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidOutputFormat.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * guanbo
+ */
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidOutputFormat.java]: same above

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSinkConfig.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package org.apache.seatunnel.connectors.seatunnel.druid.config;
+
+import java.io.Serializable;
+
+/**
+ * guanbo
+ */
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSinkConfig.java]: Check all `guanbo` places, delete it.

File: seatunnel-connectors-v2/connector-druid/pom.xml (hunk 3 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/pom.xml]: In JsonUtil OBJECT_MAPPER configure state cannot be shared.need new another ObjectMapper

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java (hunk 6 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java]: done

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidOutputFormat.java (hunk 7 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidOutputFormat.java]: done

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSinkConfig.java (hunk 8 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/config/DruidSinkConfig.java]: done

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java (hunk 5 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java]: done

File: seatunnel-connectors-v2/connector-druid/pom.xml (hunk 4 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/pom.xml]: done

File: seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java (hunk 2 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/src/main/java/org/apache/seatunnel/connectors/seatunnel/druid/client/DruidInputFormat.java:18]: done

File: seatunnel-connectors-v2/connector-druid/pom.xml (hunk 1 above)
Comment (guanboo) [reply seatunnel-connectors-v2/connector-druid/pom.xml:18]: done

**Comment by CalvinKirs** [comment]
Hi, please resolve conflicts files

**Comment by guanboo** [comment]
done
resolve conflicts files：
seatunnel-connectors-v2/pom.xml
pom.xml
plugin-mapping.properties

**Comment by guanboo** [comment]
done add source & sink doc

File: config/spark.streaming.conf.template [hunk 9]
```diff
@@ -26,7 +26,7 @@ env {
   spark.executor.instances = 2
   spark.executor.cores = 1
   spark.executor.memory = "1g"
-  spark.stream.batchDuration = 5
+  spark.streaming.batchDuration = 5
```
Comment (Hisoka-X) [inline config/spark.streaming.conf.template:29]: Please check this https://github.com/apache/incubator-seatunnel/pull/2575

File: docs/en/connector-v2/sink/Druid.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
+# Druid
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Druid.md:1]: Please update doc to match new format, reference: https://github.com/apache/incubator-seatunnel/pull/2625

File: docs/en/contribution/setup.md [hunk 11]
```diff
@@ -35,12 +35,6 @@ Otherwise, your code could not start in JetBrains IntelliJ IDEA correctly.
 ./mvnw install -Dmaven.test.skip
 ```
 
-### Building SeaTunnel from source
```
Comment (Hisoka-X) [inline docs/en/contribution/setup.md:38]: Why remove this?

File: docs/zh-CN/spark/quick-start.md [hunk 12]
```diff
@@ -34,7 +34,7 @@ ln -s seatunnel-<version> seatunnel
 ```bash
 env {
   # seatunnel defined streaming batch duration in seconds
-  spark.stream.batchDuration = 5
```
Comment (Hisoka-X) [inline docs/zh-CN/spark/quick-start.md:37]: same as above

File: pom.xml [hunk 13]
```diff
@@ -204,6 +204,11 @@
         <p3c-pmd.version>1.3.0</p3c-pmd.version>
         <maven-scm-provider-jgit.version>1.9.5</maven-scm-provider-jgit.version>
         <testcontainer.version>1.16.3</testcontainer.version>
+        <jackson-datatype-joda.version>2.12.1</jackson-datatype-joda.version>
```
Comment (Hisoka-X) [inline pom.xml:207]: Don't add the connector's dependency in root pom, just use it in connector's pom.xml, reference: https://github.com/apache/incubator-seatunnel/pull/2630/files

File: seatunnel-apis/seatunnel-api-spark/src/main/scala/org/apache/seatunnel/spark/stream/SparkStreamingExecution.scala [hunk 14]
```diff
@@ -48,12 +48,9 @@ class SparkStreamingExecution(sparkEnvironment: SparkEnvironment)
             dataset)
         }
         var ds = dataset
-
-        if (ds.take(1).length > 0) {
```
Comment (Hisoka-X) [inline seatunnel-apis/seatunnel-api-spark/src/main/scala/org/apache/seatunnel/spark/stream/SparkStreamingExecution.scala:52]: Can't delete this, check this: https://github.com/apache/incubator-seatunnel/pull/2578

**Comment by Hisoka-X** [comment]
I'm sorry, but I want reject this PR. Because too many code update but no connection with Druid Source/Sink. You can split it into multi PR, each have only one main purpose. Thanks again!

**Comment by Hisoka-X** [comment]
And you delete lots of code which shouldn't be deleted.

**Comment by Hisoka-X** [comment]
Any update you can reopen this PR again, or create another PR only with Druid Source/Sink (recommend).

**Comment by guanboo** [comment]
OK  THX！
I can create another PR only with Druid Source/Sink.Done today or tomorrow

**Comment by guanboo** [comment]
done 
#2651

**Pull request closed without merging**